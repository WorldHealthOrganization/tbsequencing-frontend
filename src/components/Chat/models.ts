export interface IUserMessage {
  email: string,
  username: string,
  firstName: string,
  lastName: string
}
