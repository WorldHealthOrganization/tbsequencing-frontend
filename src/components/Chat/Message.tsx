/** @jsxImportSource @emotion/react */
import React from 'react';
import AccountBoxIcon from '@mui/icons-material/AccountBox';
import { format, parseISO } from 'date-fns';
import {
  messageText, messageHeader, messageIcon, userSender, adminSender,
} from './styles';
import appColors from '../../styles/colors';
import H3 from '../typography/H3';

interface IMessageProps {
  sender: string,
  text: string,
  timeStamp: string,
  isOwnMessage?: boolean
}

const Message = ({
  sender, text, timeStamp, isOwnMessage,
}: IMessageProps) => (
  <div>
    <div css={messageHeader}>
      <AccountBoxIcon
        sx={messageIcon}
        htmlColor={isOwnMessage ? appColors.secondary.whoMagenta : appColors.secondary.whoBlue}
      />
      <H3 style={isOwnMessage ? userSender : adminSender}>
        {`${sender}, ${format(parseISO(timeStamp), 'HH:mm')}`}
      </H3>
    </div>
    <H3 style={messageText}>{text}</H3>
  </div>
);

export default Message;
