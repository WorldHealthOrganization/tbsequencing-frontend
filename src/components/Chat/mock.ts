import { IMessage } from '../../services/submissionApi/models';

export const mockMessages: IMessage[] = [
  {
    pk: 1,
    sender: {
      email: 'admin@email.com',
      username: 'Admin',
      firstName: '',
      lastName: '',
    },
    timestamp: '2022-09-07T13:42:16.027Z',
    content: 'please review your submission',
  },
  {
    pk: 2,
    sender: {
      email: 'delegate@email.com',
      username: 'Delegate',
      firstName: '',
      lastName: '',
    },
    timestamp: '2022-09-07T13:42:16.027Z',
    content: 'done, check one more time pls pls pls pls pls',
  },
  {
    pk: 3,
    sender: {
      email: 'admin@email.com',
      username: 'Admin',
      firstName: '',
      lastName: '',
    },
    content: 'looks good, I`m submitting',
    timestamp: '2022-09-07T13:42:16.027Z',
  },
  {
    pk: 4,
    sender: {
      email: 'delegate@email.com',
      username: 'Delegate',
      firstName: '',
      lastName: '',
    },
    content: 'Thank`s, have a nice day',
    timestamp: '2022-09-07T13:42:16.027Z',
  },
];
