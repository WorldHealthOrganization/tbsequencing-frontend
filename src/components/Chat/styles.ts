import { css } from '@emotion/react';
import { font } from '../typography/fontSizes';
import appColors from '../../styles/colors';

const chatWrapper = css({
  backgroundColor: 'rgba(221, 239, 249, 0.25)',
  borderRadius: '4px',
  height: '424px',
  marginBottom: '24px',
  overflowY: 'auto',
});

const chatHeaderContainer = css({
  marginBottom: '28px',
});

const textAreaContainer = css({
  backgroundColor: 'rgba(221, 239, 249, 0.25)',
  display: 'flex',
  flexDirection: 'column',
  section: {
    marginTop: 0,
  },
  textarea: {
    backgroundColor: 'transparent',
    border: 'none',
    fontFamily: font.regular,
    fontSize: '16px',
    '&::placeholder': {
      fontFamily: font.regular,
      fontSize: '16px',
      color: appColors.tints.whoBlue.lighter,
    },
    '&:focus': {
      outline: 'none',
    },
  },
  padding: '24px',
});

const H1Styles = css({
  color: appColors.primary.navy,
});

const H3Styles = css({
  fontWeight: 400,
  fontFamily: font.regular,
});

const noMessagesText = css({
  fontWeight: 400,
  fontFamily: font.regular,
  color: appColors.tints.whoBlue.lighter,
  textAlign: 'center',
  padding: '100px 60px 0 60px',
});

const closeIcon = {
  position: 'absolute',
  top: 15,
  right: 25,
  fill: appColors.secondary.whoBlue,
  fontSize: '30px',
  cursor: 'pointer',
};

const chatOuter = css({
  height: '100%',
  display: 'flex',
  flexDirection: 'column',
  justifyContent: 'center',
  marginLeft: '40px',
  marginRight: '24px',
});

const chatClose = css({
  cursor: 'pointer',
  border: 'none',
  backgroundColor: 'transparent',
});

const chatMessages = css({
  padding: '24px',
  marginBottom: '24px',
});

const awaitsForAnswerText = css({
  fontWeight: 400,
  color: appColors.tints.whoBlue.lighter,
  fontFamily: font.regular,
  textAlign: 'center',
});

const date = css({
  color: appColors.tints.whoBlue.lighter,
  fontWeight: 700,
  letterSpacing: '0.4px',
});

const messageText = css({
  overflowWrap: 'anywhere',
  fontFamily: font.regular,
  color: appColors.neutral.black,
  whiteSpace: 'pre-line',
  fontWeight: 400,
  marginLeft: '50px',
  marginBottom: '15px',
});

const messageHeader = css({
  display: 'flex',
  alignItems: 'center',
  marginTop: '10px',
});

const messageIcon = {
  fontSize: '40px',
  marginRight: '10px',
};

const adminSender = css({
  color: appColors.secondary.whoBlue,
  fontFamily: font.regular,
  fontWeight: 400,
});

const userSender = css({
  color: appColors.secondary.whoMagenta,
  fontFamily: font.regular,
  fontWeight: 400,
});

const messageStyles = css({
  '&:not(:last-of-type)': {
    borderBottom: `1px solid ${appColors.tints.whoBlue.lighter}`,
  },
});

const sendButton = css({
  marginLeft: 'auto',
});

export {
  chatWrapper,
  chatClose,
  messageText,
  messageHeader,
  messageIcon,
  messageStyles,
  chatMessages,
  chatOuter,
  chatHeaderContainer,
  H1Styles,
  H3Styles,
  userSender,
  adminSender,
  textAreaContainer,
  sendButton,
  noMessagesText,
  closeIcon,
  date,
  awaitsForAnswerText,
};
