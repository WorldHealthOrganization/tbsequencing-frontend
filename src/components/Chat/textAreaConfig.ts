import * as yup from 'yup';
import { validateMaxLength } from '../../utils/validators';

export interface ITextArea {
  textArea: string;
}

export const defaultValues = {
  textArea: '',
};

export const textAreaField = {
  type: 'textarea',
  name: 'textArea',
  label: '',
  placeholder: 'Start typing...',
};

export const textAreaSchema = yup.object().shape({
  textArea: validateMaxLength(2000, 'Message should contain lower than 2000 characters'),
});
