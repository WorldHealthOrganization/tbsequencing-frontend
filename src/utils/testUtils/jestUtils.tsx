import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { reducer } from '../../app/store';

const getWrapper = (
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
  } = {},
) => ({ children }: { children: ReactElement }) => (
  <Provider store={store}>{children}</Provider>
);

function render(
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
    ...renderOptions
  } = {},
) {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const getTreeJSONWithRedux = (ui: ReactElement, {
  // @ts-ignore
  preloadedState,
  store = configureStore({ reducer, preloadedState }),
} = {}) => {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  const tree = renderer
    .create(<Wrapper>{ui}</Wrapper>)
    .toJSON();

  return tree;
};

export const wrappedInRouter = (
  component: ReactElement,
) => <BrowserRouter>{component}</BrowserRouter>;

export * from '@testing-library/react';
export { render, getTreeJSONWithRedux };
