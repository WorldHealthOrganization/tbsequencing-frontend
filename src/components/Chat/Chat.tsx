/** @jsxImportSource @emotion/react */
import React, { useEffect, useState } from 'react';

import { Control, FieldErrors, useForm } from 'react-hook-form';
import CloseIcon from '@mui/icons-material/Close';
import { yupResolver } from '@hookform/resolvers/yup';
import { group, InternMap } from 'd3-array';
import { format } from 'date-fns';
import { useMsal } from '@azure/msal-react';
import {
  chatWrapper,
  chatMessages,
  chatOuter, chatHeaderContainer, H1Styles, H3Styles,
  textAreaContainer, sendButton, noMessagesText, closeIcon,
  date, messageStyles, awaitsForAnswerText,
} from './styles';

import Message from './Message';
import FormFields from '../FormFields';
import {
  defaultValues, ITextArea, textAreaField, textAreaSchema,
} from './textAreaConfig';
import H1 from '../typography/H1';
import H3 from '../typography/H3';
import { AppButton } from '../AppButton/AppButton';
import { useGetMessagesQuery, useSendMessagesMutation } from '../../services/submissionApi';
import { IMessage } from '../../services/submissionApi/models';
import { useDrawerApi } from '../../features/drawer/hooks/useDrawerApi';
import PrimaryText from '../typography/PrimaryText';
import LoadingWrapper from '../LoadingWrapper';

const groupMessagesByDate = (messages?: IMessage[]) => {
  if (!messages) {
    return new InternMap<string, IMessage[]>([]);
  }
  return group(messages, (value) => format(new Date(value.timestamp), 'PPP'));
};

export const Chat = () => {
  const { drawerContentProps, closeDrawer } = useDrawerApi();
  const { accounts } = useMsal();

  const { data, isLoading } = useGetMessagesQuery(
    drawerContentProps!.packageId as number,
    { refetchOnMountOrArgChange: true },
  );
  const [sendMessage] = useSendMessagesMutation();

  const [messages, setMessages] = useState<InternMap<string, IMessage[]>>(
    groupMessagesByDate(data),
  );

  useEffect(() => {
    setMessages(groupMessagesByDate(data));
  }, [data]);

  const {
    control, watch, formState: { errors }, getValues, setValue,
  } = useForm<ITextArea>({
    resolver: yupResolver(textAreaSchema),
    defaultValues,
    mode: 'onChange',
  });

  const watchTextarea = watch('textArea');

  const renderMessages = () => {
    if (!messages.size) {
      return (
        <H3 style={noMessagesText}>
          Leave comments for the admin
          team
        </H3>
      );
    }

    return Array.from(messages.entries()).map(([key, value], index) => {
      const isLastGroup = index === messages.size - 1;
      let shouldRenderMsgFooter = false;

      return (
        <>
          <div key={key}>
            <PrimaryText style={date}>{key}</PrimaryText>
            {value?.map((message, messageIndex) => {
              const isLastMessage = messageIndex === value.length - 1;

              if (isLastMessage && isLastGroup) {
                const isUserAuthorOfMessage = message.sender.email === accounts[0]?.username;

                shouldRenderMsgFooter = isUserAuthorOfMessage;
              }

              return (
                <div css={messageStyles} key={message.timestamp}>
                  <Message
                    sender={`${message.sender.firstName || message.sender.username} ${message.sender.lastName}`}
                    timeStamp={message.timestamp}
                    text={message.content}
                    isOwnMessage={message.sender.username === accounts[0]?.username}
                  />
                </div>
              );
            })}
          </div>
          {shouldRenderMsgFooter
            && <H3 style={awaitsForAnswerText}>The admin team will answer in a few moments</H3>}
        </>
      );
    });
  };

  const addMessage = () => {
    const newMessage = {
      pk: drawerContentProps.packageId,
      sender: {
        username: accounts[0]?.username || '',
        email: accounts[0]?.username || '',
        firstName: accounts[0]?.name || '',
        lastName: accounts[0]?.name || '',
      },
      timestamp: new Date().toISOString(),
      content: getValues().textArea,
    };

    setMessages((prevState) => {
      const newMessageKey = format(new Date(newMessage.timestamp), 'PPP');
      const newMessagesMap = new InternMap(prevState);
      const prevValues = newMessagesMap.get(newMessageKey);

      if (prevValues) {
        const newValues = [
          ...prevValues,
          newMessage,
        ];
        newMessagesMap.set(newMessageKey, newValues);
        return newMessagesMap;
      }

      newMessagesMap.set(newMessageKey, [newMessage]);
      return newMessagesMap;
    });

    sendMessage(newMessage);
    setValue('textArea', '');
  };

  const handleKeyPress = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter' && watchTextarea.trim().length && !e.shiftKey) {
      addMessage();
    }
  };

  return (
    <div css={chatOuter}>
      <CloseIcon
        onClick={closeDrawer}
        sx={closeIcon}
      />
      <div css={chatHeaderContainer}>
        <H1 style={H1Styles}>Chat</H1>
        <H3 style={H3Styles}>Share your concerns and questions with the Admin</H3>
      </div>
      <div css={chatWrapper}>
        <div css={chatMessages}>
          <LoadingWrapper centered isLoading={isLoading}>
            {renderMessages()}
          </LoadingWrapper>
        </div>
      </div>
      <div tabIndex={0} css={textAreaContainer} role="button" onKeyPress={handleKeyPress}>
        <FormFields
          control={control as unknown as Control}
          fieldArray={[textAreaField]}
          errors={errors as FieldErrors}
          loading={false}
        />
        <AppButton
          onClick={() => addMessage()}
          variant="contained"
          size="small"
          style={sendButton}
          disabled={!watchTextarea.trim().length}
        >
          Send comment
        </AppButton>
      </div>
    </div>
  );
};

export default Chat;
