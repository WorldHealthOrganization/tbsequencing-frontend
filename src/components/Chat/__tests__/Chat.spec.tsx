import React from 'react';
import { getTreeJSONWithRedux, render } from '../../../utils/testUtils/jestUtils';
import Chat from '../Chat';

describe('chat', () => {
  test('renders Chat component', () => {
    render(<Chat />);
  });

  test('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <Chat />,
    );

    expect(tree).toMatchSnapshot();
  });
});
