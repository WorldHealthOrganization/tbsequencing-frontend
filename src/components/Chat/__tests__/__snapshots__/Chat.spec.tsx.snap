// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat Should render correctly 1`] = `
<div
  className="css-vlmu6g-Chat"
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1cimjwj-MuiSvgIcon-root"
    data-testid="CloseIcon"
    focusable="false"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
    />
  </svg>
  <div
    className="css-1h90s6r-Chat"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1ing70r-MuiTypography-root-BaseText"
    >
      Chat
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-1a6atuo-MuiTypography-root-BaseText"
    >
      Share your concerns and questions with the Admin
    </p>
  </div>
  <div
    className="css-1bzwkyt-Chat"
  >
    <div
      className="css-nqsikg-Chat"
    >
      <span
        className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1ag5kwf-MuiCircularProgress-root"
        role="progressbar"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    className="css-1d1ygrd-Chat"
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <section
      className="css-4qzvqz-FormFields"
    >
      <textarea
        className="css-16gir0u-TextAreaControl"
        name="textArea"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Start typing..."
        rows={2}
        value=""
      />
    </section>
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButtonBase-root Mui-disabled css-ugn8iu-MuiButtonBase-root-MuiButton-root-AppButton"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-3cwtjx-MuiTypography-root-BaseText"
      >
        Send comment
      </p>
    </button>
  </div>
</div>
`;
